rewrite-java/src/main/java/org/openrewrite/java/JavaTemplate.java

```diff
@@ -191,11 +189,37 @@ public J visitClassDeclaration(J.ClassDeclaration classDecl, Integer p) {
                         }
                         case IMPLEMENTS: {
                             List<TypeTree> implementings = substitutions.unsubstitute(templateParser.parseImplements(substitutedTemplate));
-                            J.ClassDeclaration c = classDecl.withImplements(implementings);
+                            List<JavaType.FullyQualified> implementsTypes = implementings.stream()
+                                    .map(TypedTree::getType)
+                                    .map(TypeUtils::asFullyQualified)
+                                    .filter(Objects::nonNull)
+                                    .collect(Collectors.toList());
+                            J.ClassDeclaration c = classDecl;
+
+                            if (mode.equals(JavaCoordinates.Mode.REPLACEMENT)) {
+                                c = c.withImplements(implementings);
+                                //noinspection ConstantConditions
+                                c = c.getPadding().withImplements(c.getPadding().getImplements().withBefore(Space.format(" ")));
+                            } else {
+                                c = c.withImplements(ListUtils.concatAll(c.getImplements(), implementings));
+                            }
+                            if (c.getType() != null) {
+                                String fqn = c.getType().getFullyQualifiedName();
+                                c = c.withType(new JavaTypeVisitor<List<JavaType.FullyQualified>>() {
+                                    @Override
+                                    public JavaType visitClass(JavaType.Class aClass, List<JavaType.FullyQualified> fullyQualifieds) {
+                                        JavaType.Class c = (JavaType.Class) super.visitClass(aClass, fullyQualifieds);
+                                        if (fqn.equals(c.getFullyQualifiedName())) {
+                                            c = c.withInterfaces(ListUtils.concatAll(c.getInterfaces(), fullyQualifieds));
+                                        }
+                                        return c;
+                                    }
+                                }.visit(c.getType(), implementsTypes));
```

Comment:
Likely improved by `visitNonNull`?

Reply:
Oh, good idea, I added `JavaTypeVisitor.visitNonNull()` since it did not have this.